crates/timers/src/future.rs

```diff
@@ -71,37 +54,32 @@ impl TimeoutFuture {
     /// # Example
     ///
     /// ```no_run
-    /// # extern crate futures_rs as futures;
-    /// use futures::prelude::*;
     /// use gloo_timers::future::TimeoutFuture;
+    /// use wasm_bindgen_futures::spawn_local;
     ///
-    /// wasm_bindgen_futures::spawn_local(
-    ///     TimeoutFuture::new(1_000).map(|_| {
-    ///         // Do stuff after one second...
-    ///     })
-    /// );
+    /// spawn_local(async {
+    ///     TimeoutFuture::new(1_000).await;
+    ///     // Do stuff after one second...
+    /// });
     /// ```
     pub fn new(millis: u32) -> TimeoutFuture {
-        let mut id = None;
-        let promise = js_sys::Promise::new(&mut |resolve, _reject| {
-            id = Some(set_timeout(&resolve, millis as i32));
+        let (tx, rx) = oneshot::channel();
+        let inner = Timeout::new(millis, move || {
+            // if the receiver was dropped we do nothing.
+            let _ = tx.send(());
         });
-        debug_assert!(id.is_some());
-        let inner = JsFuture::from(promise);
-        TimeoutFuture { id, inner }
+        TimeoutFuture { inner, rx }
     }
 }
 
 impl Future for TimeoutFuture {
-    type Item = ();
-    type Error = ();
+    type Output = ();
 
-    fn poll(&mut self) -> Poll<(), ()> {
-        match self.inner.poll() {
-            Ok(Async::Ready(_)) => Ok(Async::Ready(())),
-            Ok(Async::NotReady) => Ok(Async::NotReady),
-            // We only ever `resolve` the promise, never reject it.
-            Err(_) => wasm_bindgen::throw_str("unreachable"),
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match Future::poll(Pin::new(&mut self.rx), cx) {
```

Comment:
I'd rather this be written as `Pin::new(&mut self.rx).poll(cx).map(|t| t.unwrap_throw())`